netsuitesdk/internal/client.py

```diff
@@ -533,34 +533,38 @@ def basic_stringfield_search(self, type_name, attribute, value, operator=None):
-
-    #     Usage example:
-    #         customer1 = self.Customer(externalId='customer', email='test1@example.com')
-    #         customer2 = self.Customer(externalId='another_customer', email='test2@example.com')
-    #         self.upsertList(records=[customer1, customer2])
-
-    #     :param list[CompoundValue] records: the records to be created or updated
-    #     :return: a reference to the newly created or updated records
-    #     :rtype: list[CompoundValue]
-    #     """
-
-    #     response = self.request('upsertList', record=records)
-    #     responses = response.body.writeResponse
-    #     record_refs = []
-    #     for response in responses:
-    #         status = response.status
-    #         if status.isSuccess:
-    #             record_ref = response['baseRef']
-    #             self.logger.debug('Successfully updated record of type {type}, internalId: {internalId}, externalId: {externalId}'.format(
-    #                     type=record_ref['type'], internalId=record_ref['internalId'], externalId=record_ref['externalId']))
-    #             record_refs.append(record_ref)
-    #         else:
-    #             exc = self._request_error('upsertList', detail=status['statusDetail'][0])
-    #             has_failures = True
-    #             raise exc
-    #     return record_refs
+    def upsert_list(self, records):
+        """
+        Add objects of type recordType with given externalId..
+        If a record of specified type with matching externalId already
+        exists, it is updated.
+
+        Usage example:
+            customer1 = self.Customer(externalId='customer', email='test1@example.com')
+            customer2 = self.Customer(externalId='another_customer', email='test2@example.com')
+            self.upsert_list(records=[customer1, customer2])
+
+        :param list[CompoundValue] records: the records to be created or updated
+        :return: a reference to the newly created or updated records, with a success/error status
+        :rtype: list[OrderedDict]
+        """
+
+        response = self.request('upsertList', record=records)
+        responses = response.body.writeResponseList
+        status = responses['status']
+        record_refs = []
+        if status.isSuccess:
+            for response in responses.writeResponse:
+                object_reference = response['baseRef']
+                record_ref = {
+                    "object_reference": object_reference,
+                    "status": response["status"]
+                }
+                self.logger.debug('Successfully updated record of type {type}, internalId: {internalId}, externalId: {externalId}'.format(
+                        type=object_reference['type'], internalId=object_reference['internalId'], externalId=object_reference['externalId']))
+                record_refs.append(record_ref)
+        else:
+            exc = self._request_error('upsertList', detail=status['statusDetail'][0])
+            raise exc
```

Comment:
so, this only raises _if_ the entire body is an error. 

Below is an example `writeResponseList`

It has two things: an `isSuccess` and a `list[writeResponse]`

For each `writeResponse`,  we don't raise if it's an error, just return the error message, but if the `writeResponseList` has `isSuccess="false"`, then we ought to raise because something went quite wrong and the whole write failed. In this example, you can see that both writes were failures, but the `writeResponseList` still returns success=true since there wasn't anything broken in the request

```xml
            <writeResponseList>
                <platformCore:status isSuccess="true" xmlns:platformCore="urn:core_2019_1.platform.webservices.netsuite.com"/>
                <writeResponse>
                    <platformCore:status isSuccess="false" xmlns:platformCore="urn:core_2019_1.platform.webservices.netsuite.com">
                        <platformCore:statusDetail type="ERROR">
                            <platformCore:code>INVALID_KEY_OR_REF</platformCore:code>
                            <platformCore:message>Invalid subsidiary reference key subsidiary-0.</platformCore:message>
                        </platformCore:statusDetail>
                    </platformCore:status>
                    <baseRef xsi:type="platformCore:RecordRef" type="journalEntry" externalId="c0f393ab-e221-410d-9e95-1fadb6929776" xmlns:platformCore="urn:core_2019_1.platform.webservices.netsuite.com"/>
                </writeResponse>
                <writeResponse>
                    <platformCore:status isSuccess="false" xmlns:platformCore="urn:core_2019_1.platform.webservices.netsuite.com">
                        <platformCore:statusDetail type="ERROR">
                            <platformCore:code>INVALID_KEY_OR_REF</platformCore:code>
                            <platformCore:message>Invalid subsidiary reference key subsidiary-0.</platformCore:message>
                        </platformCore:statusDetail>
                    </platformCore:status>
                    <baseRef xsi:type="platformCore:RecordRef" type="journalEntry" externalId="f5277ba6-e1b8-41b6-a4a0-c2f3a87d4f63" xmlns:platformCore="urn:core_2019_1.platform.webservices.netsuite.com"/>
                </writeResponse>
            </writeResponseList>```

Reply:
haha i had to update this so i could get some syntax highlighting here, I've been reading it in much worse states 😭 


here's a preview of how i've been seeing it ![Screen Shot 2022-06-16 at 10 53 31 AM](https://user-images.githubusercontent.com/5113256/174098167-4aced100-973f-473f-80bc-bfbc6ed4d43d.png)